dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/DefaultTrackedEntityAttributeService.java

```diff
@@ -170,64 +181,36 @@ public TrackedEntityAttribute getTrackedEntityAttribute( String uid )
     {
         Assert.notNull( trackedEntityAttribute, "tracked entity attribute is required." );
+        Assert.notNull( value, "tracked entity attribute value is required." );
 
-        if ( !trackedEntityAttribute.isUnique() || value == null )
+        if ( trackedEntityAttribute.getProgramScope() )
         {
-            return null;
+            log.error( "TrackedEntityAttribute '" + trackedEntityAttribute.getUid() + "' has a programScope. ProgramScope feature is not supported." );
+            return "Attribute " + trackedEntityAttribute.getUid() + " has a programScope. ProgramScope is not supported.";
         }
 
         TrackedEntityInstanceQueryParams params = new TrackedEntityInstanceQueryParams();
         params.addAttribute( new QueryItem( trackedEntityAttribute, QueryOperator.EQ, value, trackedEntityAttribute.getValueType(),
             trackedEntityAttribute.getAggregationType(), trackedEntityAttribute.getOptionSet() ) );
-        params.setInternalSearch( true );
 
-        if ( trackedEntityAttribute.getOrgunitScope() && trackedEntityAttribute.getProgramScope() )
-        {
-            Assert.notNull( program, "program is required for program scope" );
-            Assert.notNull( organisationUnit, "organisationUnit is required for org unit scope" );
-
-            if ( !program.getOrganisationUnits().contains( organisationUnit ) )
-            {
-                return "Organisation unit is not assigned to program " + program.getUid();
-            }
-
-            params.setProgram( program );
-            params.addOrganisationUnit( organisationUnit );
-            params.setOrganisationUnitMode( OrganisationUnitSelectionMode.SELECTED );
-        }
-        else if ( trackedEntityAttribute.getOrgunitScope() )
+        if ( trackedEntityAttribute.getOrgunitScope() )
         {
             Assert.notNull( organisationUnit, "organisation unit is required for org unit scope" );
-            params.setOrganisationUnitMode( OrganisationUnitSelectionMode.SELECTED );
             params.addOrganisationUnit( organisationUnit );
         }
-        else if ( trackedEntityAttribute.getProgramScope() )
-        {
-            Assert.notNull( program, "program is required for program scope" );
-            params.setOrganisationUnitMode( OrganisationUnitSelectionMode.ALL );
-            params.setProgram( program );
-        }
-        else
-        {
-            params.setOrganisationUnitMode( OrganisationUnitSelectionMode.ALL );
-        }
-
-        // TODO re-factor to avoid circular dependency
 
+        //to avoid circular dependency
         TrackedEntityInstanceService trackedEntityInstanceService = applicationContext.getBean( TrackedEntityInstanceService.class );
 
-        Grid instances = trackedEntityInstanceService.getTrackedEntityInstancesGrid( params );
+        Optional<TrackedEntityInstance> fetchedTei = trackedEntityInstanceService.getTrackedEntityInstanceWithUniqueAttributeValue( params );

```

Comment:
Why not call `trackedEntityInstanceStore.getTrackedEntityInstanceWithUniqueAttributeValue( params );` directly, so we avoid getting the beans from the context? This will also simplify your unit test.

Reply:
Good point. Fixed.